src/RecursiveStagesService.php

```diff
@@ -0,0 +1,187 @@
+     *
+     * @param DataObject $object
+     * @return bool
+     * @throws Exception
+     */
+    public function stagesDifferRecursive(DataObject $object): bool
+    {
+        $cacheKey = $object->getUniqueKey();
+
+        if (!array_key_exists($cacheKey, $this->stagesDifferCache)) {
+            $this->stagesDifferCache[$cacheKey] = $this->determineStagesDifferRecursive($object);
+        }
+
+        return $this->stagesDifferCache[$cacheKey];
+    }
+
+    /**
+     * Execution ownership hierarchy traversal and inspect individual models
+     *
+     * @param DataObject $object
+     * @return bool
+     * @throws Exception
+     */
+    protected function determineStagesDifferRecursive(DataObject $object): bool
+    {
+        if (!$object->exists()) {
+            // Model hasn't been saved to DB, so we can just bail out as there is nothing to inspect
+            return false;
+        }
+
+        $models = [$object];
+
+        // Compare existing content (perform full ownership traversal)
+        while ($model = array_shift($models)) {
+            if ($this->isModelDirty($model)) {
+                // Model is dirty,
+                // we can return here as there is no need to check the rest of the hierarchy
+                return true;
+            }
+
+            // Discover and add owned objects for inspection
+            $relatedObjects = $this->getOwnedObjects($model);
+            $models = array_merge($models, $relatedObjects);
+        }
+
+        // Compare deleted content,
+        // this wouldn't be covered in hierarchy traversal as deleted models are no longer present
+        $draftIdentifiers = $this->getOwnedIdentifiers($object, Versioned::DRAFT);
+        $liveIdentifiers = $this->getOwnedIdentifiers($object, Versioned::LIVE);
+
+        return $draftIdentifiers !== $liveIdentifiers;
+    }
+
+    /**
+     * Get unique identifiers for all owned objects, so we can easily compare them
+     *
+     * @param DataObject $object
+     * @param string $stage
+     * @return array
+     * @throws Exception
+     */
+    protected function getOwnedIdentifiers(DataObject $object, string $stage): array
+    {
+        $identifiers = Versioned::withVersionedMode(function () use ($object, $stage): array {
+            Versioned::set_stage($stage);
+
+            /** @var DataObject $stagedObject */
+            $stagedObject = DataObject::get_by_id($object->ClassName, $object->ID);
+
+            if ($stagedObject === null) {
+                return [];
+            }
+
+            $models = [$stagedObject];
+            $identifiers = [];
+
+            while ($model = array_shift($models)) {
+                // Compose a unique identifier, so we can easily compare models
+                // Note that we intentionally use base class here, so we can cover the situation where model class changes
+                $identifiers[] = implode('_', [$model->baseClass(), $model->ID]);
```

Comment:
```suggestion
                $identifiers[] = implode('_', [$model->baseClass(), $model->getUniqueKey()]);
```
Unless there's a reason why we wouldn't want the extra identifying data from `getUniqueKeyComponents()`, we should probably use `getUniqueKey()`.

If there is a good reason not to do that, please add a comment in the code, so we don't swap it out in the future.

Reply:
There is, will update the comment.